src/caf/core/data_structures.py

```diff
@@ -684,8 +870,564 @@ def aggregate(self, segs: list[str]):
+            a large impact when not.
+        tol: float = 1e-5
+            The RMSE between self and targets the process will attempt to reach
+            before stopping. Can also stop if consecutive iterations score the same.
+        max_iters: int = 100
+            The max number of iterations which can run before the process will stop.
+
+        Returns
+        -------
+        DVector matched to targets, rmse achieved.
+        """
+        # check DVectors compatible
+        targets = self.validate_ipf_targets(targets)
+        new_dvec = self.copy()
+        prev_rmse = np.inf
+        bypass = False
+        for i in range(max_iters):
+            for target in targets:
+                if i > 0:
+                    bypass = True
+                inner = new_dvec.copy()
+                if target.segment_translations is not None:
+                    for targ_seg, seed_seg in target.segment_translations.items():
+                        inner = inner.translate_segment(
+                            from_seg=seed_seg, to_seg=targ_seg, _bypass_validation=bypass
+                        )
+                agg = inner.aggregate(target.data.segmentation, _bypass_validation=bypass)
+                if target.zone_translation is not None:
+                    agg = agg.translate_zoning(
+                        target.data.zoning_system,
+                        trans_vector=target.zone_translation,
+                        _bypass_validation=bypass,
+                    )
+                factor = target.data.__truediv__(agg, _bypass_validation=bypass)
+                if target.zoning_diff:
+                    factor = factor.translate_zoning(
+                        self.zoning_system,
+                        trans_vector=target.zone_translation,
+                        one_to_one=True,
+                        check_totals=False,
+                    )
+                if target.segment_translations is not None:
+                    for targ_seg, seed_seg in target.segment_translations.items():
+                        factor = factor.translate_segment(
+                            from_seg=targ_seg,
+                            to_seg=seed_seg,
+                            reverse=True,
+                            _bypass_validation=bypass,
+                        )
+                new_dvec = new_dvec.__mul__(factor, _bypass_validation=bypass)
+
+            rmse = new_dvec.calc_rmse(targets)
+            print(f"RMSE = {rmse} after {i + 1} iterations.")
+            if rmse < tol:
+                print("Convergence met, returning DVector.")
+                return new_dvec
+            if abs(rmse - prev_rmse) < tol:
+                print(f"RMSE has stopped improving at {rmse}.")
+                return new_dvec
+            prev_rmse = rmse
+        warnings.warn(
+            "Convergence has not been met, and RMSE has not stopped improving. "
```

Comment:
Isn't this warning reached when max iterations is met not when RMSE has stopped improving

Reply:
Yes, that's why it says RMSE has not stopped improving

Reply:
Yes that's what this says. It hasn't converged, but RMSE is still getting better by iteration, otherwise either one of the other to conditions would have been met.